src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQuerySQLBuilder.java

```diff
@@ -0,0 +1,287 @@
+      builder.append("(");
+    }
+
+    builder.append(leftAlias).append(".").append(leftField);
+    builder.append(" = ");
+    builder.append(rightAlias).append(".").append(rightField);
+
+    if (joinOnNullKeys) {
+      builder.append(" OR (");
+      builder.append(leftAlias).append(".").append(leftField).append(" IS NULL");
+      builder.append(" AND ");
+      builder.append(rightAlias).append(".").append(rightField).append(" IS NULL");
+      builder.append(")").append(")");
+    }
+  }
+
+  protected String getBQTableName(String stageName) {
+    String result = stageToBQTableNameMap.get(stageName);
+
+    if (result == null) {
+      throw new SQLEngineException(String.format("Unable to determine BQ table name for stage '%s'", stageName));
+    }
+
+    return result;
+  }
+
+  protected void addTableNamesAndAliasesForJoinDefinition(JoinDefinition joinDefinition) {
+    for (JoinStage stage : joinDefinition.getStages()) {
+      String stageName = stage.getStageName();
+
+      if (!stageToFullTableNameMap.containsKey(stageName)) {
+        addFullTableName(stageName);
+      }
+
+      if (!stageToTableAliasMap.containsKey(stageName)) {
+        addTableAlias(stageName);
+      }
+    }
+  }
+
+  protected void addFullTableName(String stageName) {
+    String bqTableName = getBQTableName(stageName);
+    stageToFullTableNameMap.put(stageName,
+                                String.format("`%s.%s.%s`", project, dataset, bqTableName));
+  }
+
+  protected void addTableAlias(String stageName) {
+    stageToTableAliasMap.put(stageName,
+                             buildTableAlias(stageToTableAliasMap.size()));
+  }
+
+  /**
+   * Get fully qualified (quoted) table name for a supplied stage.
+   */
+  protected String getFullTableName(String stageName) {
+    String fullTableName = stageToFullTableNameMap.get(stageName);
+
+    if (fullTableName == null) {
+      throw new SQLEngineException(String.format("Unable to determine full table name for stage '%s'", stageName));
+    }
+
+    return fullTableName;
+  }
+
+  /**
+   * Get table alias for a supplied stage.
+   */
+  protected String getTableAlias(String stageName) {
+    String tableAlias = stageToTableAliasMap.get(stageName);
+
+    if (tableAlias == null) {
+      throw new SQLEngineException(String.format("Unable to determine table alias for stage '%s'", stageName));
+    }
+
+    return tableAlias;
+  }
+
+  /**
+   * Builds a column alias based on an index.
+   *
+   * The resulting Alias follows the spreadsheet column naming convention:
+   * 0 -> A
+   * ...
+   * 25 -> Z
+   * 26 -> AA
+   * 27 -> AB
+   *
+   * @param index the index to generate
+   * @return the table alias for the current index.
+   */
+  protected String buildTableAlias(int index) {
+
+    StringBuilder alias = new StringBuilder();
+
+    // Build all preceding characters in the alias, if needed.
+    if ((index / 26) > 0) {
+      alias.append(buildTableAlias((index / 26) - 1));
```

Comment:
Can you make it a for loop and not recursion? You are recursively making stringbuilders and converting them to string and appending to other stringbuilders

Reply:
Iterating means I have to "prepend" into a String, which is a N^2 operation as I have to copy the string.

I can re-factor this to not use the StringBuilder and build the string in-place.

Reply:
Re-factored in order to avoid using StringBuilder in the recursive call.